spring-data-eclipse-store/src/main/java/software/xdev/spring/data/eclipse/store/repository/lazy/SpringDataEclipseStoreLazy.java

```diff
@@ -0,0 +1,257 @@
+import java.util.Objects;
+
+import org.eclipse.serializer.reference.Lazy;
+import org.eclipse.serializer.reference.ObjectSwizzling;
+import org.eclipse.serializer.reference.Swizzling;
+
+import software.xdev.spring.data.eclipse.store.exceptions.LazyNotUnlinkableException;
+import software.xdev.spring.data.eclipse.store.repository.access.modifier.FieldAccessModifier;
+import software.xdev.spring.data.eclipse.store.repository.support.copier.working.WorkingCopier;
+
+
+/**
+ * This is the Lazy-Wrapper a user of the Spring-Data-Eclipse-Store-Library should use. Please <b>do not use the
+ * {@link Lazy}-Wrapper!</b> Because SDES is making working copies of the stored data, the {@link Lazy} does not work as
+ * expected. Instead, use this Wrapper. It brings the same functionality as the native {@link Lazy}-Wrapper but works
+ * with working copies.
+ *
+ * @param <T> the type of the lazily referenced element
+ */
+public interface SpringDataEclipseStoreLazy<T> extends Lazy<T>
+{
+	static <T> SpringDataEclipseStoreLazy.Default<T> build(final T objectToWrapInLazy)
+	{
+		return new Default<>(objectToWrapInLazy);
+	}
+	
+	SpringDataEclipseStoreLazy<T> copyWithReference();
+	
+	void unlink();
+	
+	long objectId();
+	
+	boolean isOriginalObject();
+	
+	interface Internals
+	{
+		static <T> SpringDataEclipseStoreLazy.Default<T> buildWithLazy(final Lazy<T> lazySubject)
+		{
+			return new Default<>(lazySubject);
+		}
+	}
+	
+	/**
+	 * This class is very complex and its various member variables all have their reason to exist. This code is very
+	 * difficult to read due to its the functionality explained in the {@link SpringDataEclipseStoreLazyBinaryHandler}.
+	 *
+	 * @param <T> the type of the lazily referenced element
+	 */
+	@SuppressWarnings({"java:S2065", "checkstyle:FinalClass"})
+	class Default<T> implements SpringDataEclipseStoreLazy<T>
+	{
+		private T objectToBeWrapped;
+		private Lazy<T> wrappedLazy;
+		private long objectId = Swizzling.notFoundId();
+		private transient ObjectSwizzling loader;
+		private transient WorkingCopier<T> copier;
+		private transient boolean isStored;
+		
+		private Default(final Lazy<T> lazySubject)
+		{
+			this.setWrappedLazy(lazySubject);
+		}
+		
+		private Default(final T objectToBeWrapped)
+		{
+			this.objectToBeWrapped = objectToBeWrapped;
+		}
+		
+		private Default(final long objectId, final ObjectSwizzling loader, final WorkingCopier<T> copier)
+		{
+			this.objectId = objectId;
+			this.loader = loader;
+			this.copier = copier;
+			// This object is already stored in the real storage. So it can get cleared.
+			this.setStored();
+		}
+		
+		private Lazy<T> ensureLazy()
+		{
+			if(this.wrappedLazy == null || !this.wrappedLazy.isLoaded())
+			{
+				this.wrappedLazy = this.createNewDefaultLazyWithClearableReference();
+			}
+			return this.wrappedLazy;
+		}
+		
+		@SuppressWarnings("unchecked")
+		private Lazy<T> createNewDefaultLazyWithClearableReference()
+		{
+			Objects.requireNonNull(this.loader);
```

Comment:
Maybe move those `requireNonNull` lines to constructor?

Reply:
This is not valid. In the constructor it is valid to leave the loader to null (hence the other constructors). But the method is not allowed to get executed when the loader is null.